src/Interpreters/ActionsVisitor.cpp

```diff
@@ -587,18 +590,53 @@ void ActionsMatcher::visit(const ASTFunction & node, const ASTPtr & ast, Data &
     }
 }
 
-void ActionsMatcher::visit(const ASTLiteral & literal, const ASTPtr & ast, Data & data)
+void ActionsMatcher::visit(const ASTLiteral & literal, const ASTPtr & /* ast */,
+    Data & data)
 {
-    CachedColumnName column_name;
-    if (data.hasColumn(column_name.get(ast)))
-        return;
-
     DataTypePtr type = applyVisitor(FieldToDataType(), literal.value);
+    const auto value = convertFieldToType(literal.value, *type);
+
+    // FIXME why do we have a second pass with a clean sample block over the same
+    // AST here? Anyway, do not modify the column name if it is set already.
+    if (literal.unique_column_name.empty())
+    {
+        const auto default_name = literal.getColumnName();
+        auto & block = data.getSampleBlock();
+        auto * existing_column = block.findByName(default_name);
+
+        /*
+         * To approximate CSE, bind all identical literals to a single temporary
+         * columns. We try to find the column by its default name, but after that
+         * we have to check that it contains the correct data. This might not be
+         * the case if it is a user-supplied column, or it is from under a join,
+         * etc.
+         * Overall, this is a hack around a generally poor name-based notion of
+         * column identity we currently use.
+         */
+        if (existing_column
+            && existing_column->column
+            && isColumnConst(*existing_column->column)
+            && existing_column->column->size() == 1
+            && existing_column->column->get(0) == value)
```

Comment:
For now in Field logic Null = Null. Probably we need a test for NULL constant.

Reply:
Do you mean a regression test that the NULL literal columns are collapsed?
I thought that maybe I can write a test that takes too long if the columns are not collapsed,
but no luck:
```
$ build-clang9/programs/clickhouse-client  --max_parser_depth=10000 --query "select coalesce(null, coalesce( ...repeat 1000x.... ,1 )...)"
Segmentation fault (core dumped)
```
For non-null columns, I noticed that I broke CSE because the `00984_parser_stack_overflow` failed -- it becomes really slow if you don't collapse literals. The reason for the slowness was that there is some quadratic slowdown when removing unneeded columns from block at some later stage of query execution -- didn't have time to look into that yet.

Reply:
No. I mean if some day `Null = Null` became `Null != Null` in Field logic (that could happen) we will get silent error here without functional test for NULL as constant. It would be nice to add test with constant NULL, JOIN and `isConstant()`.